lineapy/graph_reader/artifact_collection.py

```diff
@@ -42,55 +46,60 @@ def __init__(
         self.db: RelationalLineaDB = db
 
         self.input_parameters = input_parameters
+
+        # ------ Validate inputs ------
+        # check if any duplicates exist in input parameters
         if len(input_parameters) != len(set(input_parameters)):
             raise ValueError(
                 f"Duplicated input parameters detected in {input_parameters}"
             )
-
-        # Retrieve target artifact objects and group them by session ID
-        self.target_artifacts_by_session = (
-            self._get_artifacts_grouped_by_session(target_artifacts)
-        )
-
-        # Retrieve reuse precomputed artifact objects and group them by session ID
-        self.pre_computed_artifacts_by_session = (
-            self._get_artifacts_grouped_by_session(
-                reuse_pre_computed_artifacts
-            )
+        # Check if target artifacts name has been repeated in the input list
+        check_duplicates(target_artifacts)
+        # Check if entries in reuse pre computed artifacts list have been repeated
+        check_duplicates(reuse_pre_computed_artifacts)
+
+        # ------ Initialize Linea Artifacts from artifactdef ------
```

Comment:
after validation is done, the fetch from db happens without any logic involved. Batching this operation will let frameworks like sqlalchemy to optimize queries to db and get rid of overheads in connections. Note that the current level of refactor cycles through individual artifacts so a query-level optimization might not be possible, however a connection-level optimization is possible. There is potential for further gains can be had by optimizing the get_db_artifacts_from_artifactdef since we are passing the entire list. The aim here is that all these changes can be isolated to one unit-testable function rather than artifact_collection class.